### DISCUSSION TITLE
How to restore volume backup

### DISCUSSION THREAD

**Discussion by zemul** [seaweedfs/seaweedfs#4111] [Q&A]
https://github.com/seaweedfs/seaweedfs/wiki/Data-Backup

I didn't find a way to recover.

**Comment by cloudymax** [comment]
Potentially helpful: https://github.com/seaweedfs/seaweedfs/wiki/Kubernetes-Backups-and-Recovery-with-K8up

**Comment by pkk-code** [comment]
It will be great, if we have steps defined on how to restore the volume backup on non k8s.

**Comment by jameskhedley** [comment]
I have same question (need to be able to do a clean restore from archive) so I hope it's ok to expand on this a bit. Take a simple case with s3 started like this:

`weed server -master.dir=/home/s3data -dir=/home/s3metadata -filer -s3 -s3.config "settings_final.json"`

Then I make a bucket called `mybucket`, so I can see the following in weed shell:

```
> volume.list
Topology volumeSizeLimit:30000 MB hdd(volume:8/8 active:8 free:0 remote:0)
  DataCenter DefaultDataCenter hdd(volume:8/8 active:8 free:0 remote:0)
    Rack DefaultRack hdd(volume:8/8 active:8 free:0 remote:0)
      DataNode 192.168.14.127:8080 hdd(volume:8/8 active:8 free:0 remote:0)
        Disk hdd(volume:8/8 active:8 free:0 remote:0)
          volume id:5 size:8 version:3 modified_at_second:1705015768 
          volume id:6 size:8 version:3 modified_at_second:1705015768 
          volume id:7 size:8 version:3 modified_at_second:1705015768 
          volume id:8 size:478936 collection:"mybucket" file_count:171 version:3 modified_at_second:1705015865 
          volume id:1 size:64784 file_count:2 version:3 modified_at_second:1705015888 
          volume id:2 size:8 version:3 modified_at_second:1705015768 
          volume id:3 size:8 version:3 modified_at_second:1705015768 
          volume id:4 size:8 version:3 modified_at_second:1705015768 
        Disk hdd total size:543768 file_count:173 
      DataNode 192.168.14.127:8080 total size:543768 file_count:173 
    Rack DefaultRack total size:543768 file_count:173 
  DataCenter DefaultDataCenter total size:543768 file_count:173 
total size:543768 file_count:173 
```
Also on my fs:

```
$ find /home/s3data
/home/s3data
/home/s3data/vol_dir.uuid
/home/s3data/1.dat
/home/s3data/1.idx
/home/s3data/1.vif
/home/s3data/2.dat
/home/s3data/2.idx
/home/s3data/2.vif
...
/home/s3data/mybucket_8.dat
/home/s3data/mybucket_8..idx
/home/s3data/mybucket_8..vif
```

When I do `$ ./weed backup -dir=/tmp/swbackup -volumeId=8 -collection=mybucket` I get exactly the files I expect, e.g.

```
/tmp/swbackup/mybucket_8.dat
/tmp/swbackup/mybucket_8..idx
/tmp/swbackup/mybucket_8..vif
```

But, if I start a fresh seaweed s3 server and copy those files back in, nothing happens, I can't see the bucket or files, although `volume.fsck` shows the expected count as orphans, also `fs.tree` is 0:

```
> lock
> volume.fsck
total 0 directories, 0 files
dataNode:192.168.14.82:8080	volume:8	entries:178	orphan:178	100.00%	554716B
dataNode:192.168.14.82:8080	volume:3	entries:1	orphan:1	100.00%	214B

Total		entries:179	orphan:179	100.00%	554930B
This could be normal if multiple filers or no filers are used.
> fs.tree
0 directories, 0 files
```

So I gather the metadata is missing. Is it possible to rebuild metadata from the files in the volume, or would doing `fs.meta.save` and `fs.meta.load` be a good idea to try out as a way to do a clean restore?

**Comment by jameskhedley** [comment]
OK so restoring the volume files then loading the metadata back in does seem to work... sort of.

In weed shell I did the `fs.meta.save -o <filename>` then on a fresh seaweed cluster do `fs.meta.load <filename>`, as per  [the wiki](https://github.com/seaweedfs/seaweedfs/wiki/Data-Backup)

That loads the metadata and now `volume.fsck` doesn't show orphans. 

Sadly though I was getting errors trying to access the files that had been there previously via s3. Specifically, using `s3cmd` I was able to list out my files with `ls` but if I did a `get` it would fail with http 500, seaweed logs show `urls not found` error.

I think what I had to do was mount the volume, like so:

```
lock
volume.mount -volumeId 8 -node <volume_server_host>:8080
unlock
```

Which is a bit annoying to have to specify the `volumeId` and the node hostname/address but OK. Then everything works.

@pkk-code @zemul HTH

I'm all ears if there's an easier way to do restore. I think [the wiki](https://github.com/seaweedfs/seaweedfs/wiki/Data-Backup) has most of this but it requires the volume server to not be started yet, which might not suit everyone (plus it isn't that clear IMO). In my case I need to restore into a running seaweedfs.

**Comment by nithinks21** [comment]
I am also facing an issue in trying to restore the volume data.

I could take the volume data back up with weed backup command.
**_weed backup -server=localhost:9338 -dir=./$BACKUP_FOLDER/$i -volumeId=$i #-collection=<>_** 

The backed up data looks like below:
![image](https://github.com/seaweedfs/seaweedfs/assets/86155401/13898970-8c86-4cbe-a5d6-82bd567c5cb8)


Now that I have the backup of all volume IDs, I am unable to figure out which Volume ID belongs to which Volume Server.
So, to restore, I will have to copy the entire backup to all the volume server instances.

It would be great, if there is a way to find the volume id vs volume server mapping. Appreciate any inputs here.

**Comment by nkosi23** [comment]
As far as i am concerned, the restoring story makes zero sense. I have a basic setup: 1 master server with 1 volume server WITHOUT filer, but even in this case it is not clear how to restore data. Which file should i copy where? 

It is intriguing that backup is barely discussed in the docs of a storage system like seaweedfs. Being able to easily backup / restore and understand the storage format is so important, it really deserves more explaining imho. Users can't just rely on replication for backups since even EC has had a fair amount of data loss bugs recently.

If backing up/restoring is made easy, users will be able to do it frequently, and data loss bugs will be less impacting. Right now seaweedfs feels a lot like a black box with no easy way to export data. 

Me and a couple of others people at least have been scared by this situation, in light of the data loss bugs / regressions that happen from time to time. As far as i am concerned I eventually decided to go ahead  with seaweedfs but without using EC and filer-related features to stick to core stable features as much as possible, but even this use case is not totally reassuring because backing up and restoring data does not look easy.

**Reply by kmlebedev** [reply]
Backup is done quite simply. You use an external database and make a full backup of metadata using db tools once a week, a full backup is also made with deduplication of dat and idx files. If necessary, restore nearby using a ready-made pipeline, a cluster with the necessary buckets is deployed. Also, for backup copying files between full backups, I use incremental backup via filer replication/backup

**Reply by nkosi23** [reply]
Thanks a lot for your help! But I am not sure I understand, which metadata am i supposed to back up? Since I do not use Filer, I assumed that all metadata would already be including in the files produced by `weed backup`.

What I am wondering is: how to use these files to restore data in my basic setup? For example:

1.  I have backed up all my volumes using `weed backup`
2. I backup the files up to some external cloud provider for extra safety
3. The disk used by my volume server fails, so I need to buy new hardware and create a new volume server from scratch. What do I need to do? Simply copy and paste the backup files to the data directory of the volume server?
4. If the answer to 3. is "yes" and I have multiple volumes server instead of just 1, I'd do the same for every volume server needing a restore, and if i understand correctly, at the time i run `weed backup` I therefore need to keep track of which volume belongs to which volume server to ensure that I know where to restore each volume. Is that correct?
5. Last question, is there a special command that needs to be run to ensure that seaweedfs does not overrides the restored files the first time a restored volume server is started?

**Reply by chrislusf** [reply]
This needs a script or "wee shell command" to find missing volumes and copy back the volumes to the new volume servers, and then start the volume servers.

**Reply by adionditsak** [reply]
> This needs a script or "wee shell command" to find missing volumes and copy back the volumes to the new volume servers, and then start the volume servers.

Any knowledge of this being worked on? Interested in this in the case of a whole volume node goes down and becomes inconsistent with the rest of the cluster - both .dat, .idx, and .ldb. Then it would be good with a complete recover of the volume node that has replication(s) elsewhere. Eventually a way to do this as a work around by removing all the files and running fix command?

**Reply by chrislusf** [reply]
no work done yet. Welcome for PRs!

**Reply by mikhail-putilov** [reply]
> Backup is done quite simply. You use an external database and make a full backup of metadata using db tools once a week, a full backup is also made with deduplication of dat and idx files. If necessary, restore nearby using a ready-made pipeline, a cluster with the necessary buckets is deployed. Also, for backup copying files between full backups, I use incremental backup via filer replication/backup

It is super interesting if you could share with us more details of this, I am currently in the same situation. 

What I don't get is, how can I ensure that the backup of dat and idx files is correct? From Wiki page about backups, it says that I must stop all operations on SeaweedFS. I hardly can imagine a situation when I can do that (scheduled maintenance window? is it really the only option?)

How did you overcome problems with backup/restore? 

Any feedback would be appreciated.
